tests/Doctrine/Tests/ORM/Functional/Ticket/GH8499Test.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use DateTime;
+use DateTimeInterface;
+use Doctrine\DBAL\LockMode;
+use Doctrine\DBAL\Schema\SchemaException;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\GeneratedValue;
+use Doctrine\ORM\Mapping\Id;
+use Doctrine\ORM\Mapping\Table;
+use Doctrine\ORM\Mapping\Version;
+use Doctrine\ORM\OptimisticLockException;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+use function sleep;
+
+class GH8499Test extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        try {
+            $this->setUpEntitySchema(
+                [GH8499VersionableEntity::class]
+            );
+        } catch (SchemaException $e) {
+        }
+    }
+
+    /**
+     * @group GH-8499
+     */
+    public function testOptimisticLockInitializesVersionOnEntityInsert(): void
+    {
+        $entity = new GH8499VersionableEntity();
+        $entity->setName('Test Entity');
+        $entity->setDescription('Entity to test optimistic lock fix with DateTimeInterface objects');
+        $this->_em->persist($entity);
+        $this->_em->flush();
+        $this->_em->refresh($entity);
+
+        $this->assertInstanceOf(DateTimeInterface::class, $entity->getRevision(), 'Version field not set to DateTimeInterface');
+    }
+
+    /**
+     * @group GH-8499
+     */
+    public function testOptimisticLockUpdatesVersionOnEntityUpdate(): void
+    {
+        $entity = new GH8499VersionableEntity();
+        $entity->setName('Test Entity');
+        $entity->setDescription('Entity to test optimistic lock fix with DateTimeInterface objects');
+        $this->_em->persist($entity);
+        $this->_em->flush();
+        $this->_em->refresh($entity);
+        $firstVersion = $entity->getRevision();
+        sleep(1);
+
+        $entity->setName('Test Entity Updated');
+        $this->_em->persist($entity);
+        $this->_em->flush();
+        $this->_em->refresh($entity);
+        $this->assertNotEquals($firstVersion, $entity->getRevision(), 'Version field value not updated on persist');
+    }
+
+    /**
+     * @group GH-8499
+     */
+    public function testOptimisticLockWithDateTimeForVersion(): void
+    {
+        $entity = new GH8499VersionableEntity();
+        $entity->setName('Test Entity');
+        $entity->setDescription('Entity to test optimistic lock fix with DateTimeInterface objects');
+        $this->_em->persist($entity);
+        $this->_em->flush();
+
+        $firstVersion = $entity->getRevision();
+
+        $this->_em->lock($entity, LockMode::OPTIMISTIC, $firstVersion);
+        sleep(1);
+
+        $entity->setName('Test Entity Locked');
+        $this->_em->persist($entity);
+        $this->_em->flush();
+
+        $this->assertNotEquals($firstVersion, $entity->getRevision(), 'Version field value not updated on persist');
```

Comment:
We already tested that above, right? IMO, we should test the locking mechanism independently from version update. Could we achieve that by adding a setter for the version field and calling it at the right time?

Reply:
Yeah, your right, stritctly speaking I should only test two things, that the lock function works with a correct dataset and throws a exception when not. So the rest of the test cases are not needed here, I mean, the case of the version field being initailized and updated in a proper way should be tested anywhere else but not here. So, I'm going to remove the "extra" cases from the test suite and use dbal to update the record in the past as @beberlei said.

sounds good to you?

Reply:
Okie, I'm checking the OptimisticTest.php included in Tests\ORM\Functional\Locking directory again, there are a bunch of tests which covers all cases except the one with lock mechanism working as expected, the one which have detected this bug, So I'm going to write that test only. 

it's that okie?

Reply:
It's super okie 👍

Reply:
It's done, I write the test without using sleep function, following the work done in OptimisticTest.php. Check it, I think is a good approach, I followed the @beberlei advice and modified the version datetime in the past using dbal.